### DISCUSSION TITLE
Random number generator

### DISCUSSION THREAD

**Discussion by CesarChaussinand** [grame-cncm/faustlibraries#91] [Ideas]
I don't think there is a random number generation function that could be used as is by beginners. When i want to use random values as a modulation source for a synthesizer, i do it this way :
`rng = (_,_,no.noise<:!,_,!,_,!,!,!,!,_:select2(_))~_;`
I can simply use a pulse coming from a button or a ba.pulse to change the value to a random number between -1 and 1.

This function can be difficult to understand for beginners yet adding randomness to a simple synthesizer can make it way more interesting. Is it worth adding to the faust libraries ? If yes, where should i put it, noises.lib ? Is there a more efficient way to implement such function ?

**Comment by sletz** [comment]
I don't understand the` !,_,!,_,!,!,!,!,_` intermediate part ?

**Reply by CesarChaussinand** [reply]
it's to route the input of the function to the input of select2

**Reply by CesarChaussinand** [reply]
The select2 switch between the last output (when the input is 0) and a noise (when the input is 1). So each time the function get a 1, it outputs a random number and stick to it.

**Reply by CesarChaussinand** [reply]
![image](https://user-images.githubusercontent.com/44905254/130262228-bd1607c2-e12f-4de7-a672-e96825e61852.png)

**Comment by sletz** [comment]
Would using the`route` [primitive](https://faustdoc.grame.fr/manual/syntax/#route-primitive) make it simpler to read ?

**Reply by CesarChaussinand** [reply]
Indead
`rng = (_,_,no.noise:route(3,3,(1,2),(2,1),(3,3)):select2(_))~_;`

**Comment by sletz** [comment]
Why not that: `rng = (_,_,no.noise:route(3,3,(1,2),(2,1),(3,3)):select2)~_; ` ?

**Reply by CesarChaussinand** [reply]
Perfect !

**Comment by sletz** [comment]
But same as `no.noise : ba.sAndH(button("gate"));` AFAICS ?

**Reply by CesarChaussinand** [reply]
Oh ! I forgot about the sAndH function. I think it is simple enough that way, even for those who aren't familiar with the ~ connector.
Thank you.

**Comment by sletz** [comment]
Always better not to reinvent the wheel !

**Comment by josmithiii** [comment]
Also, you can use ba.latch in place of ba.sAndH to ensure one new random
number, instead of a short noise burst:

process = no.noise : ba.latch(button("Next")) : hbargraph("Value",-1,1);



On Fri, Aug 20, 2021 at 10:11 AM Stéphane Letz ***@***.***>
wrote:

> Always better not to reinvent the wheel !
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/grame-cncm/faustlibraries/discussions/91#discussioncomment-1213578>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQZKFNNMI7CL2D4SX4SE5DT52EFPANCNFSM5CQS3FSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>


-- 
Julius O. Smith III ***@***.***>
Professor of Music and, by courtesy, Electrical Engineering
CCRMA, Stanford University
http://ccrma.stanford.edu/~jos/
